fix(editor): re-add tester and prevent empty inverted diagnostics

camchenry opened the pull request:
- fixes https://github.com/oxc-project/oxc/issues/9958
- reverts + fixes https://github.com/oxc-project/oxc/pull/9982

This PR fixes an issue where the "inverted diagnostics" (i.e., the hint spans converted into diagnostic reports) cause issues in VS Code because the spans are not labeled. This happens to be the case in `no-non-null-asserted-optional-chain`: there are two diagnostic spans, but neither is labeled. This ends up creating "empty" diagnostics which cause problems. So, the fix is to simply skip creating the full diagnostic report when the message content is empty. The better fix for the future is to ensure we always label spans, as this also results in more helpful error messages.

This also re-adds the testing code which was reverted previously. I've hopefully reduced the flakiness of these tests by replacing more of the file path content to a static value.

camchenry commented:
* **#9991** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9991?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9991?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9991).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented on crates/oxc_language_server/src/linter/tester.rs:
```diff
@@ -0,0 +1,131 @@
+        Self { snapshot_suffix: None, server_linter: ServerLinter::new_with_linter(linter) }
+    }
+
+    pub fn with_snapshot_suffix(mut self, suffix: &'static str) -> Self {
```
can trhis be removed and used file names which describe the purpose of the test

camchenry replied:
I think some form of this is necessary, as it allows us to test on the same file, but with different configurations in the linter or editor. This is just a wrapper of what `insta` natively supports

Sysix commented on crates/oxc_language_server/src/linter/tester.rs:
```diff
@@ -0,0 +1,131 @@
+pub fn get_file_uri(relative_file_path: &str) -> Url {
+    let absolute_file_path = PathBuf::from(CURRENT_DIR).join(relative_file_path);
+    let file_path = format!("file://{}", absolute_file_path.display());
+    Url::parse(&file_path).unwrap()
```
is it possible to remove the `file://` prefix and use `Url::from_file_path` instead?
Did not test my suggestion

camchenry replied:
Good idea, I didn't look too closely at what other `Url` methods were available, so I didn't know about this one, thanks for the suggestion.

Sysix commented on crates/oxc_language_server/src/linter/tester.rs:
```diff
@@ -0,0 +1,131 @@
+
+use super::{error_with_position::DiagnosticReport, server_linter::ServerLinter};
+
+const CURRENT_DIR: &str = env!("CARGO_MANIFEST_DIR");
```
feels wierd to rely on `cargo` env. the `oxlint` app is using `env::current_dir()` in its `Tester`. Is it possible to do the same?

camchenry replied:
Yeah, that's fair. I had a thought that maybe this would be faster, but I'll just stick with what we do elsewhere for now.

camchenry commented:
### Merge activity

* **Mar 23, 10:16 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
fix(editor): re-add tester and prevent empty inverted diagnostics  (#9991)

- fixes https://github.com/oxc-project/oxc/issues/9958
- reverts + fixes https://github.com/oxc-project/oxc/pull/9982

This PR fixes an issue where the "inverted diagnostics" (i.e., the hint spans converted into diagnostic reports) cause issues in VS Code because the spans are not labeled. This happens to be the case in `no-non-null-asserted-optional-chain`: there are two diagnostic spans, but neither is labeled. This ends up creating "empty" diagnostics which cause problems. So, the fix is to simply skip creating the full diagnostic report when the message content is empty. The better fix for the future is to ensure we always label spans, as this also results in more helpful error messages.

This also re-adds the testing code which was reverted previously. I've hopefully reduced the flakiness of these tests by replacing more of the file path content to a static value.
